### DISCUSSION TITLE
SQLAlchemy 1.4: Warning when using IN with Select with alias

### DISCUSSION THREAD

**Discussion by Feijo** [sqlalchemy/sqlalchemy#7911] [Usage Questions] [answered]
I'm getting the following warning:

> SAWarning: Coercing Subquery object into a select() for use in IN(); please pass a select() construct explicitly

If I remove the `.alias`, the warning disappears. What's wrong with having a subquery with an alias?

Runnable example:
```python
from sqlalchemy import Column, Integer, Table, select
from sqlalchemy.orm import declarative_base

Base = declarative_base()
metadata = Base.metadata

product = Table(
    "product",
    metadata,
    Column('id', Integer, primary_key=True),
    Column('category_id', Integer),
)

# If I remove the ".alias" here, the warning disappears
subquery = select([product.c.id]).select_from(product).where(product.c.category_id.in_([1, 2, 3])).alias('product_ids')
select([product]).select_from(product).where(product.c.id.in_(subquery))
```

**Comment by CaselIT** [comment]
Hi,

That warning seems appropriate, the select inside ad IN is not a subquery. You cannot do this
```sql
SELECT *
FROM foo
WHERE id IN ((SELECT id FROM bar) as product_ids)
```

**Reply by amiltonpueljr** [reply]
Hello,

Actually SA will insert a new SELECT if you'd get an alias, the query would look like this and would run (at least on Postgres 14):

`SELECT *
FROM foo
WHERE id in (SELECT id FROM (SELECT id from bar) as product_ids)`

**Reply by zzzeek** [reply]
that's a Select object, not an Alias.   The point stays the same, in_() receives Select, not Alias.   We'd like the user to be explicit here so that SQLAlchemy isn't guessing and making assumptions.

**Reply by CaselIT** [reply]
> Actually SA will insert a new SELECT if you'd get an alias

sure, since IN does not accept an alias. The warning is exactly that: an alias was coerced to a select

**Reply by Feijo** [reply]
Alright team. I think I understand what SA is doing and WHY it's doing it. My subquery comes from a method that is used in many different places, lets say that removing that "alias" is not an option. Is there a workaround? Is there any way to "unalias" it before I put it in the `IN` statement?

**Reply by zzzeek** [reply]
call upon the ".element" member of it.  in 2.0 that will have correct type annotations and all that.

**Comment by Feijo** [comment, marked as answer]
Thanks @zzzeek for the clarification.
In summary: The IN statement should always receive a Select, not an Alias, Subquery or any other object type. You can use `.element` to access the Select object if you have an `Alias` object.
